openrtb_ext/utils.go

```diff
@@ -0,0 +1,91 @@
+package openrtb_ext
+
+import (
+    "github.com/mxmCherry/openrtb"
+    "encoding/json"
+    "math/rand"
+    "fmt"
+)
+
+// Quick little randomizer for a list of strings. Stuffing it in utils to keep other files clean
+func RandomizeList(list []BidderName) {
+    l := len(list)
+    perm := rand.Perm(l)
+    var j int
+    for i :=0; i < l; i++ {
+        j = perm[i]
+        list[i], list[j] = list[j], list[i]
+    }
+}
+
+// This will copy the openrtb BidRequest into an array of requests, where the BidRequest.Imp[].Ext field will only
+// consist of the "prebid" field and the field for the appropriate bidder parameters. We will drop all extended fields
+// beyond this context, so this will not be compatible with any other uses of the extension area. That is, this routine
+// will work, but the adapters will not see any other extension fields.
+
+// NOTE: the return map will only contain entries for bidders that both have the extension field in at least one Imp,
+// and are listed in the adapters string. wseats and bseats can be implimented by passing the bseats list as adapters,
+// or after return removing any adapters listed in wseats. Or removing all adapters in wseats from the adapters list
+// before submitting.
+
+// Take an openrtb request, and a list of bidders, and return an openrtb request sanitized for each bidder
+func CleanOpenRTBRequests(orig *openrtb.BidRequest, adapters []BidderName) (map[BidderName]*openrtb.BidRequest, []error) {
+    // This is the clean array of openrtb requests we will be returning
+    cleanReqs := map[BidderName]*openrtb.BidRequest{}
+    errList := make([]error, 0, 1)
+
+    // Decode the Imp extensions once to save time. We store the results here
+    imp_exts := make([]map[string]interface{}, len(orig.Imp))
+    // Loop over every impression in the request
+    for i := 0 ; i < len(orig.Imp) ; i++ {
+        // Unpack each set of extensions found in the Imp array
+        err := json.Unmarshal(orig.Imp[i].Ext, &imp_exts[i])
+        if err != nil {
+            return nil, []error{fmt.Errorf("Error unpacking extensions for Imp[%d]: %s", i, err.Error())}
+        }
+    }
+
+    // Loop over every adapter we want to create a clean openrtb request for.
+    for i := 0 ; i < len(adapters); i++ {
+        // Go deeper into Imp array
+        newImps := make([]openrtb.Imp, 0, len(orig.Imp))
+        bn := adapters[i].String()
+
+        // Overwrite each extension field with a cleanly built subset
+        // We are looping over every impression in the Imp array
+        for j := 0 ; j < len(orig.Imp) ; j++ {
+            // Don't do anything if the current bidder's field is not present.
+            if val, ok := imp_exts[j][bn]; ok {
+                // Start with a new, empty unpacked extention
+                newExts := map[string]interface{}{}
```

Comment:
This looks like it should reuse `adapters.ExtImpBidder` from `adapters/bidder.go`. This will help enforce the contract coupling between prebid-core & bidder implementations.

This also helps confirm that `exchange` is a better spot for this logic than `openrtb_ext`... because you can't import `adapters` from this package.

Reply:
If we want to use `adapters.ExtImpBidder` here, we need to convert the `prebid` piece into the prebid struct, and the "bidder" piece into RawJSON. Then marshal the whole thing back into a raw JSON byte array to go back into the openrtb request structure. The current implementation minimizes the conversion steps though it doesn't enforce a particular structure on the `prebid` entry.

We will be doing this per Imp and per bidder, so these transforms will be called a significant number of times.

Might be more efficient to have `map[string]openrtb.RawJSON` rather than the generic `interface{}` to explicitly stop the json module from diving deeper.

Reply:
yuck... yeah, you're right and I agree. Let's not do this.

Using `RawJSON` over `interface{}` sounds like a very good idea, though.